streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -571,68 +615,155 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
+                assignment,
+                clientMetadata,
+                partitionsForTask,
+                partitionsByHostState,
+                allOwnedPartitions,
+                activeTaskAssignments,
+                Collections.emptySet(),
+                minUserMetadataVersion);
+        }
+
+        return assignment;
+    }
+
+    private Map<String, Assignment> versionProbingAssignment(final Map<UUID, ClientMetadata> clientsMetadata,
+                                                             final Map<TaskId, Set<TopicPartition>> partitionsForTask,
+                                                             final Map<HostInfo, Set<TopicPartition>> partitionsByHostState,
+                                                             final Set<String> futureConsumers,
+                                                             final Set<TopicPartition> allOwnedPartitions,
+                                                             final int minUserMetadataVersion) {
+        final Map<String, Assignment> assignment = new HashMap<>();
 
-            final List<List<TaskId>> interleavedActive =
-                interleaveTasksByGroupId(state.activeTasks(), consumers.size());
-            final List<List<TaskId>> interleavedStandby =
-                interleaveTasksByGroupId(state.standbyTasks(), consumers.size());
+        // Since we know another rebalance will be triggered anyway, just try and generate a balanced assignment
+        // (without violating cooperative protocol) now so that on the second rebalance we can just give tasks
+        // back to their previous owners
+        // within the client, distribute tasks to its owned consumers
+        for (final ClientMetadata clientMetadata : clientsMetadata.values()) {
+            final ClientState state = clientMetadata.state;
 
-            int consumerTaskIndex = 0;
+            final Map<String, List<TaskId>> interleavedActive =
+                interleaveTasksByGroupId(state.activeTasks(), clientMetadata.consumers);
 
-            for (final String consumer : consumers) {
-                final Map<TaskId, Set<TopicPartition>> standby = new HashMap<>();
-                final List<AssignedPartition> assignedPartitions = new ArrayList<>();
+            addClientAssignments(
+                assignment,
+                clientMetadata,
+                partitionsForTask,
+                partitionsByHostState,
+                allOwnedPartitions,
+                interleavedActive,
+                futureConsumers,
+                minUserMetadataVersion);
+        }
 
-                final List<TaskId> assignedActiveList = interleavedActive.get(consumerTaskIndex);
+        return assignment;
+    }
 
-                for (final TaskId taskId : assignedActiveList) {
-                    for (final TopicPartition partition : partitionsForTask.get(taskId)) {
+    private void addClientAssignments(final Map<String, Assignment> assignment,
```

Comment:
During version probing rebalances we used to just give all clients back their old assignments, except for "future" consumers who would get an empty assignment since we can't interpret their subscriptions. But we can now determine their previous active tasks by the `ownedPartitions` so we can now put them in their assignment. 

But actually we might as well just do our best to generate a "real" (not version probing) assignment during this rebalance, so that on the second rebalance we will not have to revoke any partitions and trigger a third rebalance. So I consolidated the assignment generation into a separate method that both `computeNewAssignment` and `versionProbingAssignment` can use

Reply:
There might be kind of an ugly merge/rebasing ahead once this [version probing bugfix PR](https://github.com/apache/kafka/pull/7423) is fixed, but I think the overall behavior will remain the same